cli/src/main/java/com/devonfw/tools/ide/commandlet/CommandletFileExtractorImpl.java

```diff
@@ -0,0 +1,133 @@
+import java.nio.file.Path;
+import java.util.stream.Stream;
+
+import com.devonfw.tools.ide.cli.CliException;
+import com.devonfw.tools.ide.context.IdeContext;
+import com.devonfw.tools.ide.io.FileAccess;
+import com.devonfw.tools.ide.io.TarCompression;
+import com.devonfw.tools.ide.process.ProcessContext;
+import com.devonfw.tools.ide.tool.ToolCommandlet;
+import com.devonfw.tools.ide.util.FilenameUtil;
+
+/**
+ * Implements common {@link CommandletFileExtractor} for a {@link ToolCommandlet}
+ */
+public class CommandletFileExtractorImpl implements CommandletFileExtractor {
+
+  protected final IdeContext context;
+
+  protected final ToolCommandlet commandlet;
+
+  /**
+   * The constructor
+   * 
+   * @param context the {@link IdeContext}.
+   * @param commandlet the {@link ToolCommandlet}
+   */
+  public CommandletFileExtractorImpl(IdeContext context, ToolCommandlet commandlet) {
+
+    this.context = context;
+    this.commandlet = commandlet;
+  }
+
+  @Override
+  public void extract(Path file, Path targetDir, boolean isExtract) {
+
+    FileAccess fileAccess = this.context.getFileAccess();
+    if (isExtract) {
+      Path tmpDir = this.context.getFileAccess().createTempDir("extract-" + file.getFileName());
+      this.context.trace("Trying to extract the downloaded file {} to {} and move it to {}.", file, tmpDir, targetDir);
+      String extension = FilenameUtil.getExtension(file.getFileName().toString());
+      this.context.trace("Determined file extension {}", extension);
+      TarCompression tarCompression = TarCompression.of(extension);
+      if (tarCompression != null) {
+        fileAccess.untar(file, tmpDir, tarCompression);
+      } else if ("zip".equals(extension) || "jar".equals(extension)) {
+        fileAccess.unzip(file, tmpDir);
+      } else if ("dmg".equals(extension)) {
+        assert this.context.getSystemInfo().isMac();
+        Path mountPath = this.context.getIdeHome().resolve(IdeContext.FOLDER_UPDATES).resolve(IdeContext.FOLDER_VOLUME);
+        fileAccess.mkdirs(mountPath);
+        ProcessContext pc = this.context.newProcess();
+        pc.executable("hdiutil");
+        pc.addArgs("attach", "-quiet", "-nobrowse", "-mountpoint", mountPath, file);
+        pc.run();
+        Path appPath = fileAccess.findFirst(mountPath, p -> p.getFileName().toString().endsWith(".app"), false);
+        if (appPath == null) {
+          throw new IllegalStateException("Failed to unpack DMG as no MacOS *.app was found in file " + file);
+        }
+        fileAccess.copy(appPath, tmpDir);
+        pc.addArgs("detach", "-force", mountPath);
+        pc.run();
+      } else if ("msi".equals(extension)) {
+        this.context.newProcess().executable("msiexec").addArgs("/a", file, "/qn", "TARGETDIR=" + tmpDir).run();
+        // msiexec also creates a copy of the MSI
+        Path msiCopy = tmpDir.resolve(file.getFileName());
+        fileAccess.delete(msiCopy);
+      } else if ("pkg".equals(extension)) {
```

Comment:
What about moving each setup file type into a separate extract method e.g. extractWindowsMsi, extractLinuxPkg etc.?
Maybe create an enum for each file type too?

Reply:
Done with https://github.com/devonfw/IDEasy/pull/222/commits/c0399aa79ca64f6e2cbc36bb9fc87ae5fe73f970

Reply:
Looks much cleaner now, thanks.